src/protocol/codecs/algomsg.rs

```diff
@@ -0,0 +1,155 @@
+
+        Ok(match tag {
+            "??" => Self::UnknownMsg,
+            "AV" => Self::AgreementVote,
+            "MI" => Self::MsgOfInterest,
+            "MS" => Self::MsgDigestSkip,
+            "NP" => Self::NetPrioResponse,
+            "pi" => Self::Ping,
+            "pj" => Self::PingReply,
+            "PP" => Self::ProposalPayload,
+            "SP" => Self::StateProofSig,
+            "TS" => Self::TopicMsgResp,
+            "TX" => Self::Txn,
+            "UC" => Self::UniCatchupReq,
+            "UE" => Self::UniEnsBlockReq,
+            "VB" => Self::VoteBundle,
+            _ => return Err(invalid_data!("unexpected tag")),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Decodes a byte stream from the [input] using the [topic] codec to generate a [MsgOfInterest] message.
+    fn from(input: &mut BytesMut, topic: &mut TopicCodec) -> Result<Self, io::Error> {
+        let mut topics = topic
+            .decode(input)?
+            .ok_or_else(|| invalid_data!("couldn't decode the topics"))?;
+        if topics.len() != 1 {
+            return Err(invalid_data!("expected a single topic"));
+        }
+
+        let tag_topic = topics.pop().expect("couldn't pop the 'tags' topic");
+        if tag_topic.key.ne("tags") {
+            return Err(invalid_data!("expected 'tags' topic"));
+        }
+
+        Ok(Self {
+            tags: tag_topic.value,
+        })
+    }
+}
+
+/// [AlgoMessage] represents sets of messages received within the data field of a WebSocet message.
+///
+/// Different messages (identified by tags) use different encoding. Two encoding schemes
+/// are used:
+///  - Topic encoding (key-value string pairs)
+///  - MessagePack encoding
+#[derive(Debug)]
+pub struct AlgoMessage {
+    /// Tag represents a message type identifier.
+    pub tag: Tag,
+
+    /// Each tag has its respective payload data.
+    pub payload: Payload,
+}
+
+/// [AlgoMsgCodec] is the codec for [AlgoMessage] messages.
+#[derive(Clone)]
+pub struct AlgoMsgCodec {
+    /// The associated node's span.
+    span: Span,
+
+    /// Codec for topics which are key-value string pairs.
+    topic: TopicCodec,
+}
+
+impl AlgoMsgCodec {
+    pub fn new(span: &Span) -> Self {
+        Self {
+            topic: TopicCodec::default(),
+            span: span.clone(),
+        }
+    }
+}
+
+impl Decoder for AlgoMsgCodec {
+    type Item = AlgoMessage;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        const TAG_LEN: usize = 2;
+
+        if src.is_empty() {
```

Comment:
is this case even possible :thinking:?

Reply:
Hm, yeah. Let's remove it and use the error handling on lines 138-141